### PR TITLE
feat(stepper): Merge stepper branch into master

### PR DESCRIPTION
**Pull request by g1shin**

### CODE REVIEW

File: src/lib/stepper/_stepper-theme.scss [hunk 1]
```diff
@@ -0,0 +1,47 @@
+@import '../core/theming/palette';
+@import '../core/theming/theming';
+@import '../core/typography/_typography-utils.scss';
+
+@mixin mat-stepper-theme($theme) {
```
Comment (mmalerba) [inline src/lib/stepper/_stepper-theme.scss]: should have typography mixin too (even if it currently does nothing)

File: src/demo-app/stepper/stepper-demo.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      ])
+    });
+
+    this.nameFormGroup = this._formBuilder.group({
+      firstNameCtrl: ['', Validators.required],
+      lastNameCtrl: ['', Validators.required],
+    });
+
+    this.emailFormGroup = this._formBuilder.group({
+      emailCtrl: ['', Validators.pattern(EMAIL_REGEX)]
```
Comment (rafaelss95) [inline src/demo-app/stepper/stepper-demo.ts]: `Validators.email` can be used here.

File: src/lib/stepper/step-header.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<div [class.mat-step-icon]="icon != 'number' || selected"
+     [class.mat-step-icon-not-touched]="icon == 'number' && !selected">
+  <span *ngIf="icon == 'number'">{{index + 1}}</span>
```
Comment (rafaelss95) [inline src/lib/stepper/step-header.html]: I think `ngSwitch` fits better than 3 `*ngIf`.

```
[ngSwitch]="icon"
  *ngSwitchCase="number" ...
  *ngSwitchCase="edit"...
  *ngSwitchCase="number"...
end
```

File: src/lib/stepper/step-header.html [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  <md-icon *ngIf="icon == 'edit'">create</md-icon>
+  <md-icon *ngIf="icon == 'done'">done</md-icon>
```
Comment (rafaelss95) [inline src/lib/stepper/step-header.html]: I think `ngSwitch` fits better than 3 `*ngIf`.

```
[ngSwitch]="icon"
  *ngSwitchCase="number" ...
  *ngSwitchCase="edit"...
  *ngSwitchCase="done"...
end
```

**Comment by g1shin** [comment]
All the checks are now passing 👍 Ready for review

File: src/lib/stepper/stepper-horizontal.html [hunk 5]
```diff
@@ -0,0 +1,30 @@
+<div class="mat-horizontal-stepper-header-container">
+  <ng-container *ngFor="let step of _steps; let i = index; let isLast = last">
+    <md-step-header  class="mat-horizontal-stepper-header"
+                     (click)="step.select()"
+                     (keydown)="_onKeydown($event)"
+                     [tabIndex]="_focusIndex == i ? 0 : -1"
```
Comment (crisbeto) [inline src/lib/stepper/stepper-horizontal.html]: `==` should be `===`. Same goes for the `attr.aria-selected` and `active` bindings.

File: src/lib/stepper/stepper-vertical.html [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<div class="mat-step" *ngFor="let step of _steps; let i = index; let isLast = last">
+  <md-step-header  class="mat-vertical-stepper-header"
+                   (click)="step.select()"
+                   (keydown)="_onKeydown($event)"
+                   [tabIndex]="_focusIndex == i ? 0 : -1"
```
Comment (crisbeto) [inline src/lib/stepper/stepper-vertical.html]: Use strict equals here and the `attr.aria-selected` and `active` bindings.

File: src/lib/stepper/stepper-vertical.html [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                   [active]="step.completed || selectedIndex == i"
+                   [optional]="step.optional">
+  </md-step-header>
+
+  <div class="mat-vertical-content-container" [class.mat-stepper-vertical-line]="!isLast">
+    <div class="mat-vertical-stepper-content" role="tabpanel"
+         [@stepTransition]="_getAnimationDirection(i)"
+         [id]="_getStepContentId(i)"
+         [attr.aria-labelledby]="_getStepLabelId(i)"
+         [attr.aria-expanded]="selectedIndex == i">
```
Comment (crisbeto) [inline src/lib/stepper/stepper-vertical.html]: This should also use strict equals.

File: src/lib/stepper/stepper-vertical.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  styleUrls: ['stepper.css'],
+  inputs: ['selectedIndex'],
+  host: {
+    'class': 'mat-stepper-vertical',
+    'role': 'tablist',
+  },
+  animations: [
+      trigger('stepTransition', [
+        state('previous', style({height: '0px', visibility: 'hidden'})),
+        state('next', style({height: '0px', visibility: 'hidden'})),
```
Comment (crisbeto) [inline src/lib/stepper/stepper-vertical.ts]: Since the `previous` and `next` states have the same styling, you can combine them:

```ts
state('next, previous', style({height: '0px', visibility: 'hidden'}))
```

File: src/lib/stepper/stepper.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+`md-horizontal-stepper` selector can be used to create a horizontal stepper, and
+`md-vertical-stepper` can be used to create a vertical stepper. `md-step` components need to be
+placed inside either one of the two stepper components.
+
+### Labels
+If a step's label is only text, then the `label` attribute can be used.
+```html
+<md-vertical-stepper>
+  <md-step label="Step 1">
+    Content 1
```
Comment (crisbeto) [inline src/lib/stepper/stepper.md]: Consider using some real-world text for the example, instead of `Content X`.

File: src/lib/stepper/stepper.spec.ts [hunk 10]
```diff
@@ -0,0 +1,767 @@
...
+import {MdHorizontalStepper} from './stepper-horizontal';
+import {MdVerticalStepper} from './stepper-vertical';
+import {NoopAnimationsModule} from '@angular/platform-browser/animations';
+import {FormControl, FormGroup, ReactiveFormsModule, Validators} from '@angular/forms';
+import {MdStepperNext, MdStepperPrevious} from './stepper-button';
+import {dispatchKeyboardEvent} from '@angular/cdk/testing';
+import {ENTER, LEFT_ARROW, RIGHT_ARROW, SPACE} from '@angular/cdk/keycodes';
+import {MdStepper} from './stepper';
+
+const EMAIL_REGEX = /^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
```
Comment (crisbeto) [inline src/lib/stepper/stepper.spec.ts]: Since the unit test doesn't really care about what you're validating, you can use a simpler regex that only validates a certain value, e.g. `const UNICORN_REGEX = /unicorn/`.

File: src/lib/stepper/stepper.spec.ts [hunk 11]
```diff
@@ -0,0 +1,767 @@
...
+      assertPreviousStepperButtonClick(stepperComponent, fixture);
+    });
+
+    it('should set the correct step position for animation', () => {
+      assertCorrectStepPosition(stepperComponent, fixture);
+    });
+
+    it('should support keyboard events to move and select focus', () => {
+      let stepHeaders = fixture.debugElement.queryAll(By.css('.mat-horizontal-stepper-header'));
+      assertCorrectKeyboardInteraction(stepperComponent, fixture, stepHeaders);
```
Comment (crisbeto) [inline src/lib/stepper/stepper.spec.ts]: Not a fan of having the assertions be in separate functions, because it makes failures harder to track down, although in this case the assertions seem pretty complicated. Thoughts @jelbourn?

File: src/lib/stepper/stepper.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  /** Content for step label given by <ng-template matStepLabel> or <ng-template mdStepLabel>. */
+  @ContentChild(MdStepLabel) stepLabel: MdStepLabel;
+
+  /** Original ErrorStateMatcher that checks the validity of form control. */
+  private _originalErrorStateMatcher: ErrorStateMatcher;
+
+  constructor(mdStepper: MdStepper,
+              @Optional() @SkipSelf() @Inject(MD_ERROR_GLOBAL_OPTIONS) errorOptions: ErrorOptions) {
+    super(mdStepper);
+    this._originalErrorStateMatcher =
```
Comment (crisbeto) [inline src/lib/stepper/stepper.ts]: The ternary with a nested || is a little hard to follow. It might be cleaner if it used a regular `if/else`.

File: src/cdk/stepper/stepper.ts [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    this._stepper.selected = this;
+  }
+}
+
+@Directive({
+  selector: '[cdkStepper]',
+})
+export class CdkStepper {
+  /** The list of step components that the stepper is holding. */
+  @ContentChildren(CdkStep) _steps: QueryList<CdkStep>;
```
Comment (mmalerba) [inline src/cdk/stepper/stepper.ts]: Anything that's a `@ContentChild(ren)` or `@ViewChild(ren)` as well as anything referenced in the template must be public, they start with `_` to indicate they're internal

**Comment by mmalerba** [comment]
@g1shin please address @rafaelss95's and @crisbeto's stylistic comments in a follow-up PR. The RTL issue that @crisbeto pointed out is a bug and should be fixed in this PR before merging

**Comment by mmalerba** [comment]
fixed RTL bugs
